### PR TITLE
Feature/snapshot clustering

### PR DESCRIPTION
**Pull request by ClaraBuettner**
In this branch, snapshot clustering uses p_max_pu * p_nom instead of p_set of generators for clustering.
Additionally, users can choose which constraint they would like to add for storage units in snapshot clustering, either 'daily_bounds' or 'soc_constraint'. Using 'soc_constraint' leads to better results of storage dispatch but needs more time than 'daily_bounds'.

### CODE REVIEW

**Comment by ClaraBuettner** [comment]
Snapshot clustering is adjusted in other branches, too (e.g. https://github.com/openego/eTraGo/tree/feature/sclopf). So I think it might be easier to merge this all together at once.

**Comment by ulfmueller** [comment]
anyhow, if it works fine - why not have this function in the release?!

**Comment by ClaraBuettner** [comment]
it's true, we can add this to this release. Since seasonal storage constraints were added by other users too I wanted to compare the implementation first. But changes would still be possible, so we can add this now.

File: etrago/cluster/snapshot.py [hunk 1]
```diff
@@ -68,30 +83,44 @@ def tsam_cluster(timeseries_df, typical_periods=10, how='daily'):
...
-    start = []
+    start=[]
     # get the first hour of the clusterCenterIndices (days start with 0)
     for i in clusterCenterIndices:
-        start.append(i * hours)
+        start.append(i*hours)
 
     # get a list with all hours belonging to the clusterCenterIndices
-    nrhours = []
+    nrhours=[]
```
Comment (ulfmueller) [inline etrago/cluster/snapshot.py:123]: Were these style changes intentional?

File: etrago/cluster/snapshot.py [hunk 2]
```diff
@@ -103,7 +132,42 @@ def tsam_cluster(timeseries_df, typical_periods=10, how='daily'):
...
+    for i in range(1,int(x)):
+        j=1
+        while j <= hours: 
+            nr_day.append(i)
+            j=j+1
+    df_i_h = pd.DataFrame({'Timeseries': timeseries_df.index,
+                        'Candidate_day': nr_day})
+    df_i_h.set_index('Timeseries',inplace=True)
+
+    return df_cluster, cluster_weights, dates, hours, df_i_h
```
Comment (ulfmueller) [inline etrago/cluster/snapshot.py:170]: the doubling of the return statement looks very weird to me. Intentional?

File: etrago/cluster/snapshot.py (hunk 1 above)
Comment (ClaraBuettner) [reply etrago/cluster/snapshot.py:123]: No, they were not. I can't remember why I changed this, it's not pep8-conform. I can re-change that

File: etrago/cluster/snapshot.py (hunk 2 above)
Comment (ClaraBuettner) [reply etrago/cluster/snapshot.py:170]: I don't see another return statement (the one in 106 is deleted and a new one added in 170), can you help me?

**Comment by ClaraBuettner** [comment]
You can choose between 'daily bounds' and 'soc_constraints' here: 
https://github.com/openego/eTraGo/blob/6470fc6b7badabce1635e4ecbe8c4da84768bdbf/etrago/appl.py#L539-L543
Maybe it's better to move it to the args, I can change this.

File: etrago/cluster/snapshot.py (hunk 1 above)
Comment (ulfmueller) [reply etrago/cluster/snapshot.py:123]: that would be nice.

Comment (ClaraBuettner) [reply etrago/cluster/snapshot.py]: I did this in my latest commit. Looking at the old commit it only effected three lines, or did you see other issues?

File: etrago/cluster/snapshot.py [hunk 3]
```diff
@@ -112,9 +176,13 @@ def run(network, n_clusters=None, how='daily',
...
 
     # calculate clusters
-    tsam_ts, cluster_weights, dates, hours = tsam_cluster(
-        prepare_pypsa_timeseries(network), typical_periods=n_clusters,
-        how=how)
+    df_cluster, cluster_weights, dates, hours, df_i_h= tsam_cluster(
+                prepare_pypsa_timeseries(network),
+                typical_periods=n_clusters,
+                how='daily',
+                extremePeriodMethod = 'None')
```
[the remark this answers is not in the document]
Comment (ClaraBuettner) [reply etrago/cluster/snapshot.py:183]: I remember this error, it is solved in the feature/sclopf-branch. I'll check if I can easily copy the fix to this branch

[the remark this answers is not in the document]
Comment (ClaraBuettner) [reply etrago/cluster/snapshot.py:189]: This problem is fixed in my latest commit

**Pull request merged**